---
title: "ChIPseq read processing and mapping"
author: "Kärt Mätlik"
date: "2024-10-10"
output: html_document
---

- test picard and macs3

# Load libraries
```{r}
library(Rfastp)
library(Rsubread)
library(Rsamtools)
library(tidyverse)
library(magrittr)
library(rtracklayer)
```

# Set up files & directories
```{r}
read_1 <- c("R1_001.fastq.gz", "R1_002.fastq.gz", "R1_003.fastq.gz")
outputdir <- "C:/Users/kvare/Documents/Seq/ChIPseq/Exp1/"
outputdir_wsl <- "/mnt/c/Users/kvare/Documents/Seq/ChIPseq/Exp1/" #This is required to run picard and macs3 through wsl. Note that the path starts with /mnt/c/ instead of C:/ 
samplelist <- c("sample1", "sample2", "sample3")
inputlist <- c("sample3", "sample3", "sample3") #List of input samples for each sample. Must be in sample list as well.
ref_genome <- "C:/Users/kvare/Documents/Seq/RefGenome/mm10_mainchrs"
```

# Quality control & trimming
```{r}
for(i in 1:length(samplelist)){
  qc_report <- rfastp(
    read1 = read_1[i],
    outputFastq = paste0(outputdir, samplelist[i])
  )
}
```

# Mapping to reference genome
```{r}
for(i in 1:length(samplelist)){
  
  # Align to reference genome
  mapping_report <- align(index = ref_genome,
                             readfile1 = paste0(outputdir, samplelist[i], "_R1.fastq.gz"),
                             output_format = "BAM",
                             output_file = paste0(outputdir, samplelist[i], ".bam"), 
                             type = "dna",
                             phredOffset = 33, 
                             nthreads = 8,
                             unique = TRUE)
  mapping_report %>% write.table(paste0(outputdir, samplelist[i], "_mapping_summary.csv"))
  
  # Sort and index BAM
  sortBam(file = paste0(outputdir, samplelist[i], ".bam"), 
          destination = paste0(outputdir, samplelist[i], "_sorted"))
  
  # Remove unsorted bam files to save space
  file.remove(paste0(outputdir, samplelist[i], ".bam"))
}
```

# Run picard to remove duplicates
```{r}
#Picard does not work on Windows, so I have installed it on WSL2. Before running MarkDuplicates, we need to add read group ID and other read groups to the bam file, because picard does not work without them. Unless you want to concatenate multiple bam files, it isn't very important what you set these values. See https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups for details

#For wsl commands to run, it seems we need to include the absolute path to picard.jar from /home.

#Run picard AddOrReplaceReadGroups to add read groups. This is required until I figure out how to add read groups with Rsubread readGroup ID.
for(i in 1:length(samplelist)){
  picard_command <- paste0("wsl java -jar /home/kart/picard/build/libs/picard.jar AddOrReplaceReadGroups -I ",
                           outputdir_wsl, samplelist[i], "_sorted.bam",
                           " -O ",
                           outputdir_wsl, samplelist[i], "_withRG.bam",
                           " -ID 1 -LB lib1 -PL illumina -PU unit1 -SM 1")
  system(picard_command)
}

#With only one sample run:
#system("wsl java -jar /home/kart/picard/build/libs/picard.jar AddOrReplaceReadGroups -I /mnt/c/Users/kvare/Documents/Seq/ATAC/EzNd_CRB_7m_20241011/Ezh1WT_EzNd54_sorted.bam -O /mnt/c/Users/kvare/Documents/Seq/ATAC/EzNd_CRB_7m_20241011/Ezh1WT_EzNd54_withRG.bam -ID 1 -LB lib1 -PL illumina -PU unit1 -SM 1")

#Remove extra bam files to save space
for(i in 1:length(samplelist)){
  file.remove(paste0(outputdir, samplelist[i], "_sorted.bam"))
}

#picard MarkDuplicates
for(i in 1:length(samplelist)){
  picard_command <- paste0("wsl java -jar /home/kart/picard/build/libs/picard.jar MarkDuplicates -I ",
                           outputdir_wsl, samplelist[i], "_withRG.bam",
                           " -O ",
                           outputdir_wsl, samplelist[i], "_deduplicated.bam",
                           " -M ",
                           outputdir_wsl, samplelist[i], "_metrix.csv, --REMOVE_DUPLICATES true")
  system(macs_command)
}

#With only one sample run:
#system("wsl java -jar /home/kart/picard/build/libs/picard.jar MarkDuplicates -I /mnt/c/Users/kvare/Documents/Seq/ATAC/EzNd_CRB_7m_20241011/Ezh1WT_EzNd54_withRG.bam -O /mnt/c/Users/kvare/Documents/Seq/ATAC/EzNd_CRB_7m_20241011/Ezh1WT_EzNd54_deduplicated.bam -M /mnt/c/Users/kvare/Documents/Seq/ATAC/EzNd_CRB_7m_20241011/Ezh1WT_EzNd54_metrix.csv --REMOVE_DUPLICATES true")


#Remove extra bam files to save space
for(i in 1:length(samplelist)){
  file.remove(paste0(outputdir, samplelist[i], "_withRG.bam"))
}
```

# Index deduplicated BAMs
```{r}
for(i in 1:length(samplelist)){
  indexBam(paste0(outputdir, samplelist[i], "_deduplicated.bam"))
}
```

# Create bigwigs
```{r}
for(i in 1:length(samplelist)){
  # Mapping summary over chromosomes
  mapped_stats <- idxstatsBam(paste0(outputdir, samplelist[i], "_deduplicated.bam"))
  mapped_stats %>% write_csv(paste0(outputdir, samplelist[i], "_mapped_stats.csv"))
  
  # Create bigwig files
  read_coverage <- coverage(paste0(outputdir, samplelist[i], "_deduplicated.bam"),
                            weight = (10^6)/sum(mapped_stats[,"mapped"]))
  export.bw(read_coverage, paste0(outputdir, samplelist[i], ".bw"))
}
```

# Peak calling with macs3 (single read)
```{r}
#macs3 does not work on Windows, so I have installed it to WSL2. We need to provide the absolute path to macs3 (can check this by running which macs3 on Linux terminal). Also you may need to add macs3 to your PATH (I don't know if this is actually required). 

#This is the code you would run directly from Linux terminal. 
#macs3 callpeak -t data/sample1_deduplicated.bam -n sample1 -f BAMPE --outdir data/macs3_callpeak -g mm 

#Run macs3 callpeak
for(i in 1:length(samplelist)){
  macs_command <- paste0(c("wsl /home/kart/miniconda3/bin/macs3 callpeak -t "), 
                       outputdir_wsl, samplelist[i], "_deduplicated.bam",
                       " -c ",
                       outputdir_wsl, inputlist[i], "_deduplicated.bam",
                       " -n ",
                       samplelist[i],
                       " -f BAM --outdir ",
                       outputdir_wsl,
                       " -g mm")
  system(macs_command)
}

#With only one sample run:
#system("wsl /home/kart/miniconda3/bin/macs3 callpeak -t /mnt/c/Users/kvare/Documents/Seq/ATAC/EzNd_CRB_7m_20241011/Ezh1WT_EzNd74_deduplicated.bam -n Ezh1WT_EzNd74 -f BAMPE --outdir /mnt/c/Users/kvare/Documents/Seq/ATAC/EzNd_CRB_7m_20241011/ -g mm)
```

